Simplenote/NoticePresenter.swift

```diff
@@ -0,0 +1,170 @@
+import UIKit
+
+class NoticePresenter: KeyboardObservable {
+
+    // MARK: Properties
+    //
+    private var containerView = PassthruView(frame: .zero)
+    var noticeView: NoticeView?
+    private var noticeVariableConstraint: NSLayoutConstraint?
+
+    var isPresenting: Bool {
+        noticeView != nil
+    }
+    private var keyboardHeight: CGFloat = .zero
+    private var keyboardFloats: Bool = false
+    private var keyboardNotificationTokens: [Any]?
+    var timer: Timer?
+
+
+    // MARK: Lifecycle
+    //
+    deinit {
+        stopListeningToKeyboardNotifications()
+    }
+
+    // MARK: Presenting/Dismissing Methods
+    //
+    func presentNoticeView(_ noticeView: NoticeView, completion: @escaping () -> Void) {
+        guard let keyWindow = getKeyWindow() else {
+            return
+        }
+        self.noticeView = noticeView
+
+        containerView.addSubview(noticeView)
+        keyWindow.addFillingSubview(containerView)
+
+        noticeVariableConstraint = noticeView.topAnchor.constraint(equalTo: containerView.bottomAnchor)
+        noticeVariableConstraint?.isActive = true
+        noticeView.centerXAnchor.constraint(equalTo: containerView.centerXAnchor).isActive = true
+        containerView.layoutIfNeeded()
+
+        displayNotificationView() {
+            completion()
+        }
+    }
+
+    private func displayNotificationView(completion: @escaping () -> Void) {
+        prepareConstraintToDisplayNotice()
+
+        UIView.animate(withDuration: UIKitConstants.animationLongDuration) {
+            self.containerView.layoutIfNeeded()
+        } completion: { (_) in
+            completion()
+        }
+    }
+
+    private func makeBottomConstraintConstant() -> CGFloat {
+        let constant = Constants.bottomMarginConstant
+
+        if keyboardFloats || keyboardHeight == .zero {
+            return constant
+        }
+
+        return constant - keyboardHeight
+    }
+
+    private func prepareConstraintToDisplayNotice() {
+        noticeVariableConstraint?.isActive = false
+
+        let constant = makeBottomConstraintConstant()
+        noticeVariableConstraint = noticeView?.bottomAnchor.constraint(equalTo: containerView.bottomAnchor, constant: constant)
+        noticeVariableConstraint?.isActive = true
+    }
+
+    func dismissNotification(completion: @escaping () -> Void) {
+        guard let noticeView = noticeView else {
+            return
+        }
+
+        let delay = noticeView.handler == nil ? UIKitConstants.animationDelayExtraLong : UIKitConstants.animationDelayExtraExtraLong
```

Comment:
Perhaps it'd be clear to move those constants as a local `NoticePresenter` thing? (since it's not really an animation delay, but a "Dismissal timeout" or so?